File: front/src/services/device-registration.service.ts
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { environment } from 'src/environments/environment';
import { AirConditionerRegistrationDTO, AmbientSensorRegistrationDTO, EVChargerRegistrationDTO, HomeBatteryRegistrationDTO, IrrigationSystemRegistrationDTO, LampRegistrationDTO, SolarPanelRegistrationDTO, VehicleGateRegistrationDTO, WashingMachineRegistrationDTO } from 'src/model/device-registration';


@Injectable({
  providedIn: 'root'
})
export class DeviceRegistrationService {

  constructor(private http:HttpClient) { }

  registerAmbientSensor(ambientSensor: AmbientSensorRegistrationDTO): Observable<any> {
    console.log(environment.apiHost);
    return this.http.post<any>(environment.apiHost + "/device-registration/ambient-sensor", ambientSensor, {withCredentials: true});
  }

  registerAC(ac: AirConditionerRegistrationDTO): Observable<any> {
    console.log(environment.apiHost);
    return this.http.post<any>(environment.apiHost + "/device-registration/ac", ac, {withCredentials: true});
  }

  registerWashingMachine(wm: WashingMachineRegistrationDTO): Observable<any> {
    console.log(environment.apiHost);
    return this.http.post<any>(environment.apiHost + "/device-registration/washing-machine", wm, {withCredentials: true});
  }

  registerSolarPanel(panel: SolarPanelRegistrationDTO): Observable<any> {
    console.log(environment.apiHost);
    return this.http.post<any>(environment.apiHost + "/device-registration/solar-panel", panel, {withCredentials: true});
  }

  registerLamp(lamp: LampRegistrationDTO): Observable<any> {
    console.log(environment.apiHost);
    return this.http.post<any>(environment.apiHost + "/device-registration/lamp", lamp, {withCredentials: true});
  }

  registerVehicleGate(gate: VehicleGateRegistrationDTO): Observable<any> {
    console.log(environment.apiHost);
    return this.http.post<any>(environment.apiHost + "/device-registration/vehicle-gate", gate, {withCredentials: true});
  }

  registerHomeBattery(battery: HomeBatteryRegistrationDTO): Observable<any> {
    console.log(environment.apiHost);
    return this.http.post<any>(environment.apiHost + "/device-registration/home-battery", battery , {withCredentials: true});
  }

  registerIrrigationSystem(irrigationSystem: IrrigationSystemRegistrationDTO): Observable<any> {
    console.log(environment.apiHost);
    return this.http.post<any>(environment.apiHost + "/device-registration/irrigation-system", irrigationSystem, {withCredentials: true});
  }

  registerEVCharger(charger: EVChargerRegistrationDTO): Observable<any> {
    console.log(environment.apiHost);
    return this.http.post<any>(environment.apiHost + "/device-registration/evcharger", charger, {withCredentials: true});
  }
}  

export interface ResponseMessageDTO {
  message: string

}

File: nvt_back/nvt_back/Services/DeviceRegistrationService.cs
using nvt_back.DTOs;
using nvt_back.DTOs.DeviceRegistration;
using nvt_back.Model.Devices;
using nvt_back.Repositories.Interfaces;
using nvt_back.Services.Interfaces;

namespace nvt_back.Services
{
    public class DeviceRegistrationService : IDeviceRegistrationService
    {
        private readonly IDeviceRegistrationRepository _deviceRegistrationRepository;
        private readonly IImageService _imageService;

        public DeviceRegistrationService(IDeviceRegistrationRepository deviceRegistrationRepository,
            IImageService imageService)
        {
            _deviceRegistrationRepository = deviceRegistrationRepository;
            _imageService = imageService;
        }

        public void Add(DeviceRegistrationDTO dto)
        {
            Device deviceForDb = null;
            if (dto is AirConditionerRegistrationDTO)
                deviceForDb = new AirConditioner((AirConditionerRegistrationDTO)dto);
            else if (dto is AmbientSensorRegistrationDTO)
                deviceForDb = new AmbientSensor((AmbientSensorRegistrationDTO)dto);
            else if (dto is EVChargerRegistrationDTO)
                deviceForDb = new EVCharger((EVChargerRegistrationDTO)dto);
            else if (dto is HomeBatteryRegistrationDTO)
                deviceForDb = new HomeBattery((HomeBatteryRegistrationDTO)dto);
            else if (dto is IrrigationSystemRegistrationDTO)
                deviceForDb = new IrrigationSystem((IrrigationSystemRegistrationDTO)dto);
            else if (dto is LampRegistrationDTO)
                deviceForDb = new Lamp((LampRegistrationDTO)dto);
            else if (dto is SolarPanelRegistrationDTO)
                deviceForDb = new SolarPanel((SolarPanelRegistrationDTO)dto);
            else if (dto is VehicleGateRegistrationDTO)
                deviceForDb = new VehicleGate((VehicleGateRegistrationDTO)dto);
            else if (dto is WashingMachineRegistrationDTO)
                deviceForDb = new WashingMachine((WashingMachineRegistrationDTO)dto);

            string filePath = this._imageService.SaveImage(dto.Image);
            deviceForDb.Image = filePath;
            this._deviceRegistrationRepository.Add(deviceForDb);
        }
    }
}
